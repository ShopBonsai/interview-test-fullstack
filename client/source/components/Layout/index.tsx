import * as React from "react";
import ProductsList from "../Products";
import Toolbar from "./Toolbar";

const Layout: React.FC = () => {
  return (
    <>
      <Toolbar />
      <ProductsList />
    </>
  );
};

export default Layout;
