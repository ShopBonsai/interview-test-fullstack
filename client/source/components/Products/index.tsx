import React from "react";
import {
  CardTitle,
  CardSubtitle,
  CardText,
  Button,
  CardBody,
  Media,
} from "reactstrap";
import "./styles.css";
import { Spinner } from "../UI/Spinner";
import { IProduct } from "../../../../interfaces/IProduct";
import { timerEnd, timerStart } from "../../utils/getServerLoad";
import { useProducts } from "../../context/products-context";
import { useCart } from "../../context/shoppingcart-context";

const ProductsList: React.FC = () => {
  const {
    filteredProducts,
    loading,
    error,
    checkStock,
    updateStock,
  } = useProducts();

  const { addToCart } = useCart();

  const onClickHandler = (id: string) => {
    if (checkStock(id)) {
      const product = updateStock(id);
      if (product) {
        addToCart(product);
      }
    }
  };

  if (loading) {
    timerStart();
    return <Spinner />;
  }

  if (error || !filteredProducts) {
    timerEnd();
    return (
      <div>
        <h3>Could not fetch data!</h3>
      </div>
    );
  }

  if (filteredProducts.length < 1) {
    timerEnd();
    return (
      <div>
        <h3>No products available</h3>
      </div>
    );
  }

  timerEnd();
  return (
    <div className="product-list">
      {filteredProducts.map((product: IProduct) => {
        const {
          color,
          description,
          quantity,
          image,
          name,
          price,
          size,
        } = product;
        return quantity > 0 ? (
          <Media key={product.id} className="product-card">
            <Media left href="#">
              <Media object src={image} alt="Product image cap" />
            </Media>
            <CardBody>
              <CardTitle style={{ fontWeight: 600 }}>{name}</CardTitle>
              <CardTitle>Price: {price}</CardTitle>
              <CardSubtitle>Color: {color}</CardSubtitle>
              <CardSubtitle>Size: {size}</CardSubtitle>
              <CardSubtitle>Stock: {quantity}</CardSubtitle>
              <CardText>Details: {description}</CardText>
              <Button
                color="primary"
                size="lg"
                block
                onClick={() => onClickHandler(product.id)}
              >
                Add to Cart
              </Button>
            </CardBody>
          </Media>
        ) : null;
      })}
    </div>
  );
};

export default ProductsList;
