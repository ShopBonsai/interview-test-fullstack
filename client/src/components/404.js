import React from 'react';
import { Link } from "react-router-dom";

const ErrorPage = (props) =>{
  return (
    <div className="error-page">
      <div className="error-page__content">
        <h1>Oops!</h1>
        <p>The page you’re looking for can’t be found. Go back to your previous page or use the search bar to try again.</p>

        <div className="actions">
          <Link to="/store" className="text-primary">Go to Store</Link>
        </div>
      </div>
    </div>
  );
}

export default ErrorPage;