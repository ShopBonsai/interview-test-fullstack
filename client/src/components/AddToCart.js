import React from 'react';
import { Form, Button, Spinner } from 'reactstrap';
import { Mutation } from 'react-apollo';
import { gql } from 'apollo-boost';

const ADD_TO_CART = gql`
	mutation addToCart($id: String!, $quantity: Int!) {
		addToCart(product: { id: $id, quantity: $quantity }) {
			success
			message
			totalNumberOfCartItems
		}
	}
`;

const AddToCart = ({ children, productId, quantity, updateCart }) => {
	const handleUpdateCart = (res) => {
		// TODO: check if not successful, implemement error toast/message
		if (res.data.addToCart.success) {
			updateCart(res.data.addToCart.totalNumberOfCartItems);
		}
	};

	return (
		<Form>
			{children}
			<Mutation
				mutation={ADD_TO_CART}
				variables={{ id: productId, quantity: quantity }}
			>
				{(AddToCart, { loading, error }) => {
					// TODO: implement better error handling
					if (error) console.log(error);

					return (
						<Button
							color='primary'
							size='lg'
							block
							onClick={() => AddToCart().then((res) => handleUpdateCart(res))}
							disabled={loading}
							style={{
								padding: loading ? '.30rem 1rem' : '.5rem 1rem',
							}}
						>
							{loading ? <Spinner color='light' /> : 'Add to cart'}
						</Button>
					);
				}}
			</Mutation>
		</Form>
	);
};

export default AddToCart;
