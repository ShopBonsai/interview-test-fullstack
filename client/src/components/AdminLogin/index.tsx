import * as React from "react";
import { gql, useMutation } from "@apollo/client";

import Form from "./Form";
import { isLoggedInVar } from "../../apollo/variables";
import "./styles.css";

export const LOGIN_ADMIN = gql`
  mutation LoginAdmin($email: String!, $password: String!) {
    loginAdmin(email: $email, password: $password) {
      email
    }
  }
`;

export default function Login() {
  const [login, { error }] = useMutation<any, any>(LOGIN_ADMIN, {
    onCompleted() {
      isLoggedInVar(true);
    },
  });

  if (error) {
    return <p>An error occurred</p>;
  }

  return (
    <div className="admin-login">
      <Form login={login} />
    </div>
  );
}
