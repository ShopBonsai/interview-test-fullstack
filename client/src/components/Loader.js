import React from 'react';
import { Spinner } from 'reactstrap';

const styles = {
    container: {
        height: '100vh',
        width: '100vw',
        display: 'flex',
    },
    loader: {
        height: '3rem',
        width: '3rem',
    },
    text: {
        marginBottom: '1rem',
    },
    tile: {
        boxShadow: '5px 5px 15px rgb(136 10 142 / 40%)',
        borderRadius: '10px',
        display: 'flex',
        flexDirection: 'column',
        margin: 'auto',
        alignItems: 'center',
        padding: '2rem',
    },
};

const Container = ({ children }) => {
    return <div style={styles.container}>{children}</div>;
};

export const Loader = () => {
    return (
        <div>
            <Container>
                <div style={styles.tile}>
                    <h3 style={styles.text}>Loading products...</h3>
                    <Spinner style={styles.loader} color="primary" />
                </div>
            </Container>
        </div>
    );
};
