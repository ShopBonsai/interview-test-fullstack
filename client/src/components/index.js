export { Login } from "./Login";
export { NavHeader } from "./Nav";
export { ProductsList } from "./Products";
export { Signup } from "./Signup";
