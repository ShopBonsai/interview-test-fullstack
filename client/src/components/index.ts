export { default as Products } from './Products';
export { default as Nav } from './Nav';
export { default as Cart } from './Cart';
