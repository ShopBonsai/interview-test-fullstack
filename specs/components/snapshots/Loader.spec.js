import React from 'react';
import renderer from 'react-test-renderer';

import { Loader } from '../../../client/src/components/Loader';

describe('Snapshot', ()=> {
    it('should render correctly', () => {
       const tree = renderer.create(<Loader/>).toJSON();
       
        expect(tree).toMatchSnapshot();
    });
});