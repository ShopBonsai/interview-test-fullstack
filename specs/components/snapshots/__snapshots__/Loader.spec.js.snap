// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot should render correctly 1`] = `
<div>
  <div
    style={
      Object {
        "display": "flex",
        "height": "100vh",
        "width": "100vw",
      }
    }
  >
    <div
      style={
        Object {
          "alignItems": "center",
          "borderRadius": "10px",
          "boxShadow": "5px 5px 15px rgb(136 10 142 / 40%)",
          "display": "flex",
          "flexDirection": "column",
          "margin": "auto",
          "padding": "2rem",
        }
      }
    >
      <h3
        style={
          Object {
            "marginBottom": "1rem",
          }
        }
      >
        Loading products...
      </h3>
      <div
        className="spinner-border text-primary"
        role="status"
        style={
          Object {
            "height": "3rem",
            "width": "3rem",
          }
        }
      >
        <span
          className="sr-only"
        >
          Loading...
        </span>
      </div>
    </div>
  </div>
</div>
`;
